include/ensmallen_bits/lookahead/lookahead_impl.hpp

```diff
@@ -0,0 +1,191 @@
+
+// In case it hasn't been included yet.
+#include "lookahead.hpp"
+
+#include <ensmallen_bits/function.hpp>
+
+namespace ens {
+
+template<typename BaseOptimizerType, typename DecayPolicyType>
+inline Lookahead<BaseOptimizerType, DecayPolicyType>::Lookahead(
+    const BaseOptimizerType& baseOptimizer,
+    const double stepSize,
+    const size_t k,
+    const size_t maxIterations,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy,
+    const bool exactObjective) :
+    baseOptimizer(baseOptimizer),
+    stepSize(stepSize),
+    k(k),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy),
+    exactObjective(exactObjective),
+    isInitialized(false)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename BaseOptimizerType, typename DecayPolicyType>
+template<typename DecomposableFunctionType,
+         typename MatType,
+         typename GradType,
+         typename... CallbackTypes>
+typename std::enable_if<IsArmaType<GradType>::value,
+typename MatType::elem_type>::type
+Lookahead<BaseOptimizerType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    MatType& iterateIn,
+    CallbackTypes&&... callbacks)
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+  typedef typename MatTypeTraits<MatType>::BaseMatType BaseMatType;
+  typedef typename MatTypeTraits<GradType>::BaseMatType BaseGradType;
+
+  typedef Function<DecomposableFunctionType, BaseMatType, BaseGradType>
+      FullFunctionType;
+  FullFunctionType& f(static_cast<FullFunctionType&>(function));
+
+  // The decay policy internally use a templated class so that
+  // we can know MatType and GradType only when Optimize() is called.
+  typedef typename DecayPolicyType::template Policy<BaseMatType, BaseGradType>
+      InstDecayPolicyType;
+
+  // Make sure we have all the methods that we need.
+  traits::CheckDecomposableFunctionTypeAPI<FullFunctionType, BaseMatType,
+      BaseGradType>();
+  RequireFloatingPointType<BaseMatType>();
+  RequireFloatingPointType<BaseGradType>();
+  RequireSameInternalTypes<BaseMatType, BaseGradType>();
+
+  BaseMatType& iterate = (BaseMatType&) iterateIn;
+
+  // Check if the optimizer implements HasMaxIterations() and override the
+  // parameter with k.
+  if (callbacks::traits::HasMaxIterationsSignature<BaseOptimizerType>::value)
+  {
+    baseOptimizer.MaxIterations() = k;
+  }
+  else
+  {
+    Warn << "The BaseOptimizerType does not have a definition of "
+        << "MaxIterations(), don't override the number of iteration.";
+  }
+
+  // Check if the optimizer implements ResetPolicy() and override the reset
+  // policy setting if true.
+  if (callbacks::traits::HasResetPolicySignature<BaseOptimizerType>::value &&
+      baseOptimizer.ResetPolicy())
+  {
+    Warn << "Parameters are reset before every Optimize call; set "
+        << "ResetPolicy() to false.";
+    baseOptimizer.ResetPolicy() = false;
```

Comment:
Maybe it's worth adding a boolean parameter to `Lookahead` that will set whether or not `ResetPolicy()` is called at the start of every outer iteration?  It's not clear to me what they do in the paper implementation, though, which is probably what we should choose as the default if we offer an option like this.

Reply:
Good idea.